### DISCUSSION TITLE
RateLimit-Reset Header : getting time to wait for refill even limit is not exceeded

### DISCUSSION THREAD

**Discussion by xavier-b** [vladimir-bukhtoyarov/bucket4j#197] [Q&A] [answered]
Hi,

It would be interesting to get remaining time for a refill, even if the limit is not reached. This allows client to not exceed the limit.
Today, we have to exceed the limit to know how many time before refill.

Indeed, `ConsumptionProbe.getNanosToWaitForRefill()` return zero if limit not exceeded.
https://github.com/vladimir-bukhtoyarov/bucket4j/blob/8a6333b2653ae17da6d32211260d4968a03cea78/bucket4j-core/src/main/java/io/github/bucket4j/ConsumptionProbe.java#L104-L111

Extract of [RateLimit Header Fields for HTTP draft-polli-ratelimit-headers-00](https://tools.ietf.org/id/draft-polli-ratelimit-headers-00.html#rfc.section.3.3) :

> A client SHOULD NOT exceed the quota-units expressed in RateLimit-Remaining before the time-window expressed in RateLimit-Reset.

Actually, with bucket4j, it's not possible to have RateLimit-Reset without exceed the limit. Isn't it ?

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
@xavier-b 

Hello, ```ConsumptionProbe#getNanosToWaitForRefill``` has nothing in common with ```RateLimit-Reset```. As mentioned in the javadocs, it returns the time required to refill **requested** amount of tokens, not the time required to refill tokens up to max possible values.

To get such information you need to use Verbose API:
```java
VerboseResult<ConsumptionProbe> result = bucket.asVerbose().tryConsumeAndReturnRemaining(1);
ConsumptionProbe probe = result.getValue();
Diagnostics diagnostics = result.getDiagnostics();
if (probe.isConsumed()) {
    long nanosToRefillUpToLimit = diagnostics.calculateFullRefillingTime();
    ...
} else {
    long nanosToRefillUpToLimit = diagnostics.calculateFullRefillingTime();
    long nanosToRefillRequestedTokens = probe.getNanosToWaitForRefill();
    ...
}
```
See this doc for more details https://bucket4j.com/7.0.0/toc.html#verbose-api especially this https://bucket4j.com/7.0.0/toc.html#example-of-verbose-api-usage

**Reply by xavier-b** [reply]
Thanks,
That works fine. 
That makes more sense to use it with "intervally" refill.
When do you expect to be able to release 6.4.1 ?

**Reply by vladimir-bukhtoyarov** [reply]
6.4.1 was released last week https://mvnrepository.com/artifact/com.github.vladimir-bukhtoyarov/bucket4j-core/6.4.1
